compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -1004,31 +1000,66 @@ private void reportMissingRecordFieldDiagnostics(List<RecordLiteralNode.RecordFi
         if (bType.tag == TypeTags.UNION) {
             Set<BType> expTypes = ((BUnionType) bType).getMemberTypes();
 
-            List<BType> possibleTypes =
-                    expTypes.stream()
-                            .filter(type -> type.tag == TypeTags.MAP ||
-                                    (type.tag == TypeTags.RECORD &&
-                                            (!((BRecordType) type).sealed ||
-                                                    isCompatibleClosedRecordLiteral((BRecordType) type,
-                                                            recordLiteral))))
-                            .collect(Collectors.toList());
+            List<BType> possibleTypes = new ArrayList<>();
```

Comment:
Looking at the algorithm I think instead of an array list this could be HashSet. Feel like we are iterating over ```possibleTypes``` several times for searching. For loop and contains.

Reply:
IMO, possible types will often have only around 2-3 types. Given that, I'm wondering if there'll be a significant gain switching to a set?